lib/request-proxy/index.js

```diff
@@ -230,6 +256,75 @@ proto.handleRequest = function handleRequest(head, body, cb) {
     }
 };
 
+RequestProxy.prototype._decrementInflightRequests = function _decrementInflightRequests() {
+    this.numInflightRequests--;
+    if (this.numInflightRequests < 0) {
+        // We miscounted number of inflight requests. Correct by setting to min.
+        this.ringpop.stat('increment', 'requestProxy.miscount.decrement');
+        this.numInflightRequests = 0;
+    }
+
+    this.ringpop.stat('gauge', 'requestProxy.inflight',
+        this.numInflightRequests);
+};
+
+RequestProxy.prototype._incrementInflightRequests = function _incrementInflightRequests() {
+    this.numInflightRequests++;
+    var max = this.ringpop.config.get('maxInflightRequests');
+    if (this.numInflightRequests > max) {
+        // We miscounted number of inflight requests. Correct by setting to max.
+        this.ringpop.stat('increment', 'requestProxy.miscount.increment');
+        this.numInflightRequests = max;
+    }
+
+    this.ringpop.stat('gauge', 'requestProxy.inflight',
+        this.numInflightRequests);
+};
+
+RequestProxy.prototype._hasExceededMaxInflightRequests =
+    function _hasExceededMaxInflightRequests() {
+    var config = this.ringpop.config;
+    var backpressureEnabled = config.get('backpressureEnabled');
+    var max = config.get('maxInflightRequests');
+
+    // We add one to the current number of inflight requests to account for
+    // the current request for which we are evaluating against the maximum.
+    return backpressureEnabled && (this.numInflightRequests + 1) > max;
+};
+
+RequestProxy.prototype._hasExceededMaxEventLoopLag =
+    function _hasExceededMaxEventLoopLag() {
+    var config = this.ringpop.config;
+    var backpressureEnabled = config.get('backpressureEnabled');
+    var max = config.get('maxEventLoopLag');
+    return backpressureEnabled && this.ringpop.lagSampler.currentLag > max;
+};
+
+RequestProxy.prototype._shouldApplyBackpressure =
+    function _shouldApplyBackpressure(egressOrIngress, callback) {
+    if (this._hasExceededMaxInflightRequests()) {
+        this.ringpop.stat('increment', 'requestProxy.refused.inflight');
+        callback(new TooManyRequestsError({
```

Comment:
you're probably right. let me revisit the flow and make sure nextTick is/isn't appropriate here.